docs/design/multi_slo_doc_support.md

```diff
@@ -0,0 +1,102 @@
+
+## Problem
+
+### Overview
+
+Today, these tools are programmed to assume, within the context of app-interface data, that 'service' and 'slo-document' have a one-to-one or one-to-zero relationship (i.e. every service can have exactly 0 or 1 slo-documents).
+
+This is consistant with most services today, but some services do in fact have multiple SLO documents
+
+We want to properly enable the usage pattern of defining more-than-one SLO documents for any given service. We want to support 'service' and 'slo-document' having a one-to-N relationship (where N can be zero or any positive integer).
+
+### Examples
+
+The service 'cincinnati' [has **a single** SLO-document](https://gitlab.cee.redhat.com/service/app-interface/-/tree/1b368998a913be3f2d45a13934eb6403d4e751de/data/services/cincinnati/slo-documents). It has [three SLOs](https://gitlab.cee.redhat.com/service/app-interface/-/blob/1b368998a913be3f2d45a13934eb6403d4e751de/data/services/cincinnati/slo-documents/cincinnati.yml#L13-47) defined. These can be seen correctly formatted in its [report YML doc](https://gitlab.cee.redhat.com/service/app-interface/-/blob/fae9d83da7db25c2ec5364eb4a6ebe8ee542e845/data/reports/Cincinnati/2021-08-01.yml#L63-78) and its [visual report page](https://visual-app-interface.devshift.net/reports#/reports/Cincinnati/2021-08-01.yml). **This is working as intended.**
+
+The service 'ocm' [has **two** SLO-documents](https://gitlab.cee.redhat.com/service/app-interface/-/tree/1b368998a913be3f2d45a13934eb6403d4e751de/data/services/ocm/slo-documents). Each of these documents define SLOs with the same set of names (latency, errors, availability). There are a total of 6 SLOs defined for this serice (2 docs * 3 SLOs-per-doc = 6; 2 SLOs named 'latency', 2 SLOs named 'errors', and 2 SLOs named 'availability'). However, the 'ocm' [visual report page](https://visual-app-interface.devshift.net/reports#/reports/ocm/2021-08-01.yml) only renders 3 SLOs. **This is NOT working as intended.** The tools do not take into account the possibility of a service defining the same SLO name multiple times across multiple docs for the same service. Presumably the SLO datapoints in dash.db are incorrectly overwriting one another or are incorrectly being aggregated together.
+
+The service ['service-registry'](https://gitlab.cee.redhat.com/service/app-interface/-/tree/6762ce9ae3c34abedc21abf768437a5e7753652a/data/services/service-registry/slo-documents) would have [a similar problem](https://visual-app-interface.devshift.net/reports#/reports/service-registry/2021-08-01.yml)(although its visual report page fails to render for a seemingly-unrelated-reason) to 'ocm'.
+
+## Proposal
+
+### Overview
+
+The 'name' ([example](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L7)) of these SLO docs will be introduced as an identifier for the SLO metric data stored in dashdotdb.
```

Comment:
Just a warning here on the API. Introducing a new field here would make the API non-backward compatible if the field is mandatory in queries (esp. GET I guess).
Either we accept that because we control all the API clients, or we need to make this backward compatible in some ways: allow queries without SLODoc and return as today, fail if several record match (from several SLODoc)

Reply:
The name linked here is the name of the service.  Do you want the name of the slo doc?  That would be in the list of slos in the app-interface schema

Reply:
> That would be in the list of slos in the app-interface schema

@rrati 

Can you link an example of what you mean?

Are you referring to [this](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L16)?

Reply:
https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L13

Reply:
@rrati 

Perhaps I am severely misunderstanding something.

You're saying [this](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L7) isn't the "name" of an [SLO doc](https://gitlab.cee.redhat.com/service/app-interface/-/blob/15d1ce504a9233a04d6119e8d57e58a3bcc31cb0/graphql-schemas/schema.yml#L1916)?

Isn't what you just linked the name of the `SLO` (defined [here](https://gitlab.cee.redhat.com/service/app-interface/-/blob/15d1ce504a9233a04d6119e8d57e58a3bcc31cb0/graphql-schemas/schema.yml#L1904)) and **not** the name of the `SLO doc` (defined [here](https://gitlab.cee.redhat.com/service/app-interface/-/blob/15d1ce504a9233a04d6119e8d57e58a3bcc31cb0/graphql-schemas/schema.yml#L1920))?

My understanding is that [this](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/accounts-mgmt.yml#L13)(your link) and [this](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/services/ocm/slo-documents/clusters-service.yml#L13) use identical SLO names for the same service (ocm)... which is problematic for the way [reports are produced for this service](https://gitlab.cee.redhat.com/service/app-interface/-/blob/8bba50902109207d7e8a0b8f856bec92ede1e482/data/reports/ocm/2021-08-01.yml#L128-143).

Reply:
I think you're right.  I linked to the name of the SLO vs the name of the slo document.  My bad.

Reply:
Discussion has shifted to: https://github.com/app-sre/dashdotdb/pull/51#discussion_r701909921